### PR TITLE
UI: Add exchange of one-time token on application load

### PR DESCRIPTION
**Pull request by backspace**
This adds UI support for receiving the one-time token passed via query parameter, as in #10134
and related PRs, and exchanging it for its corresponding secret ID. When this works, it’s mostly
invisible, as seen here when used with `nomad ui -authenticate`, where the OTT is briefly onscreen:

authenticate

You can also see the amended message when authentication fails suggesting to `-authenticate`.

When it fails, it shows a whole-page error:

failed

Unfortunately, I was unable to produce the desired UX of having the OTT be erased from the URL
after processing, so I’m submitting this for review with that sad caveat. This will only be seen when
jumping directly with an identifier (`nomad ui -authenticate jobname`), as `nomad ui -authenticate`
forwards to `/ui/?ott=…` which then forwards to `/ui/jobs` after the exchange is complete, but it’s
undesirable regardless. I’ll continue trying to find a way to cleanly and reliably remove it.

leftover

The whole-page error currently has this minimal text:

> Failed to exchange the one-time token.

This is an edge case where I can’t actually imagine what would produce it, but the error doesn’t
really help you if it does happen… I’m open to suggestions for something better!

### CODE REVIEW

**Comment by tgross** [comment]
@backspace just a heads up that https://github.com/hashicorp/nomad/pull/10134 has been merged.

**Comment by backspace** [comment]
Thanks! Sadly I likely won’t be able to return to this until March 22, but hopefully things proceed smoothly then 😯

File: ui/app/templates/application.hbs [hunk 1]
```diff
@@ -10,6 +10,11 @@
         <p data-test-error-message class="subtitle">
           The cluster has no leader. <a href="https://www.nomadproject.io/guides/outage.html"> Read about Outage Recovery.</a>
         </p>
+      {{else if this.isOTTExchange}}
+        <h1 data-test-error-title class="title is-spaced">Token Exchange Error</h1>
+        <p data-test-error-message class="subtitle">
+          Failed to exchange the one-time token.
+        </p>
```
Comment (backspace) [inline ui/app/templates/application.hbs:17]: I followed the existing `no-leader-error` pattern and created the `ott-exchange-error` that’s checked for here to display this custom message… maybe a more generic known error type with custom title and message would be useful eventually 🤔

File: ui/app/routes/application.js [hunk 2]
```diff
@@ -34,31 +49,31 @@ export default class ApplicationRoute extends Route {
...
+      (queryParam && queryParam !== currentRegion) ||
+      (!queryParam && currentRegion !== defaultRegion)
+    ) {
+      this.system.reset();
+      this.store.unloadAll();
+    }
+
+    this.set('system.activeRegion', queryParam || defaultRegion);
+
+    return promises;
```
Comment (DingoEatingFuzz) [inline ui/app/routes/application.js:76]: Yesss, more `await` 🤩

File: ui/app/services/token.js [hunk 3]
```diff
@@ -42,6 +42,14 @@ export default class TokenService extends Service {
   })
   fetchSelfToken;
 
+  @task(function*(oneTimeToken) {
+    const TokenAdapter = getOwner(this).lookup('adapter:token');
+
+    const token = yield TokenAdapter.exchangeOneTimeToken(oneTimeToken);
+    this.secret = token.secret;
+  })
+  exchangeOneTimeToken;
```
Comment (DingoEatingFuzz) [inline ui/app/services/token.js:51]: ⚠️ totally random thoughts that you don't have to take seriously

Hm, even though it's consistent to the file to use an EC task here, looking at this with fresh eyes it seems like a simple promise would suffice? We aren't doing any fancy chaining or error control flow. Idk.

File: ui/app/styles/components/empty-message.scss [hunk 4]
```diff
@@ -22,4 +22,20 @@
...
+
+  .terminal-container {
+    display: flex;
+    justify-content: center;
+    margin-top: 1.25rem;
+  }
+
+  .terminal {
+    background: $grey-lighter;
+    border-radius: 2px;
```
Comment (DingoEatingFuzz) [inline ui/app/styles/components/empty-message.scss:34]: You should use [`$radius`](https://github.com/hashicorp/nomad/blob/main/ui/app/styles/core/variables.scss#L18) here.

File: ui/app/styles/components/empty-message.scss [hunk 5]
```diff
@@ -22,4 +22,20 @@
...
+    padding: 0.75rem 1rem;
+
+    .prompt {
+      color: $grey;
+    }
```
Comment (DingoEatingFuzz) [inline ui/app/styles/components/empty-message.scss:39]: Thank you for the attention to detail 🥳

File: ui/app/templates/application.hbs (hunk 1 above)
Comment (DingoEatingFuzz) [reply ui/app/templates/application.hbs:17]: Yeah. Errors and notifications I think are high on the "to refactor" list.

File: ui/app/templates/components/forbidden-message.hbs [hunk 6]
```diff
@@ -17,6 +17,12 @@
         requisite
       {{/if}}
       permission to view this.
+      <br>
+      <br>
```
Comment (DingoEatingFuzz) [inline ui/app/templates/components/forbidden-message.hbs:21]: Woah, is this the first instance of two paragraphs of text in the UI? I couldn't find any prior art for this type of spacing.

What do you think about changing `empty-message-body` to being a `div` and then making each line a `p` and then creating this paragraph spacing with margins instead?

Something like this in the `.empty-message-body` selector:

```scss
p:not(:last-child) {
  margin-bottom: 0.5rem; // idk how much spacing this should actually be, somewhere between 0.5 and 1.
}
```

File: ui/app/routes/application.js [hunk 7]
```diff
@@ -25,7 +26,21 @@ export default class ApplicationRoute extends Route {
...
 
-  beforeModel(transition) {
+  async beforeModel(transition) {
+    let exchangeOneTimeToken;
+
+    if (transition.to.queryParams.ott) {
+      exchangeOneTimeToken = this.get('token.exchangeOneTimeToken')
+        .perform(transition.to.queryParams.ott)
+        .catch(() => {
+          this.controllerFor('application').set('error', new OTTExchangeError());
```
Comment (DingoEatingFuzz) [inline ui/app/routes/application.js:36]: I think ideally this error casting would happen in the data layer. Presumably in the token adapter.

Not a big deal though if that ends up being more trouble than it's worth.

**Review by DingoEatingFuzz — CHANGES_REQUESTED** [review]
🎉 Works as advertised!

I agree that it would be nice if the query param always got deleted, but I think your choice to ship this as is and follow up with that improvement is the right call.

Everything here is looking good aside from the couple `br` elements I'd like to avoid introducing. Aside from that, comments are commentary and weak suggestions.

File: ui/app/services/token.js (hunk 3 above)
Comment (backspace) [reply ui/app/services/token.js]: I took it seriously in a6e6b94, it makes sense to remove the needless wrapping, thank you!

File: ui/app/styles/components/empty-message.scss (hunk 4 above)
Comment (backspace) [reply ui/app/styles/components/empty-message.scss]: Done in 7b319fc, thanks 😀

File: ui/app/templates/components/forbidden-message.hbs (hunk 6 above)
Comment (backspace) [reply ui/app/templates/components/forbidden-message.hbs]: I did find [this two-paragraph](https://github.com/hashicorp/nomad/blob/a82697a2f024555cd68d0fa4ba5b52257d885e6b/ui/app/templates/components/fs/file.hbs#L31-L37) empty message, though it’s slightly different in that the second paragraph is a button. In 412da49 I added this within `.empty-message-body`:

```
    &:not(:last-child) {
      margin-bottom: 1rem;
    }
 ```

and removed the unsightly `<br>`s, thanks for the push there. Here are screenshots of these two two-paragraph messages, does it seem suitable to you?

double-p-existing

double-p-terminal

File: ui/app/routes/application.js (hunk 7 above)
Comment (backspace) [reply ui/app/routes/application.js]: It was easy enough, makes sense, done in d5e2e5d, thank you!

File: ui/app/templates/components/forbidden-message.hbs (hunk 6 above)
Comment (DingoEatingFuzz) [reply ui/app/templates/components/forbidden-message.hbs]: 💯

**Review by DingoEatingFuzz — APPROVED** [review]
:shipit: Nice work!

**Pull request merged**